CO-32: Versioning and Release Strategy

carlpulley opened the pull request:
- [x] Add in code for publishing plugin to Sonatype
- [x] Add in code for (general) releasing to Artifactory/Nexus
- [x] Single project testing
- [ ] ~~Multi-module project testing~~ Using git tags to trigger release does not allow multi-module projects to have independent versioning on a per module basis - hence this sub-task is out of scope for the current work
- [x] Update README with new functionality
- [x] Update README with versioning and release actions for the plugin
- [x] Deal with library evictions
- [x] Ensure code successfully builds in Jenkins
- [x] Jenkins integration documented
- [x] Removal of the `sbt-sensible` plugin should be possible with the merging of this code
- [x] Update versioning strategy in light of architectural review comments
---
* code for generating Jira release notes will be handled as part of https://cakesolutions.jira.com/browse/CO-97
---
- [x] Ready for review
- [x] Ready for merge

Commit:
CO-32: Versioning and Release Strategy

Commit:
Added in release notes task stub

fommil commented on README.md:
```diff
@@ -187,6 +187,20 @@ Configuration of this plugin should be avoided in local project SBT build files.
 
+## `CakePublishPlugin`: Artifact Publishing
+
+Plugin requirements: `SonatypePlugin` and `PgpPlugin`
```
is this for publishing the `sbt-cake` itself? I was under the impression that we wouldn't want to have artifact/sonatype publishing in typical proprietary projects. It's probably best not to include it by default to avoid accidental mistakes.

fommil commented on build.sbt:
```diff
@@ -35,6 +35,7 @@ scalacOptions in Compile ++= Seq("-feature", "-deprecation")
 addSbtPlugin("com.dwijnand" % "sbt-dynver" % "1.2.0")
 addSbtPlugin("com.eed3si9n" % "sbt-buildinfo" % "0.7.0")
 addSbtPlugin("com.jsuereth" % "sbt-pgp" % "1.0.1")
+addSbtPlugin("org.xerial.sbt" % "sbt-sonatype" % "2.0")
```
I really don't think we should go this far... typical cake projects are **not** published on sonatype and this just opens up the possibility of accidental error.

carlpulley replied:
That's a good point. I'll rethink what we need to do here.

Commit:
Refactored with:
- Sonatype publishing for plugin itself
- Generic Artifactory/Nexus publishing for plugin client code
- Releasing based on dynamic versioning

fommil commented:
the "release strategy" used to be

```
git tag -s v1.1.1
sbt publishSigned sonatypeRelease
git push --tags
```

After CO-32 I have no idea how to make a release... why would you want to add all that extra boilerplate? Nor do I understand why we've added all this boilerplate to `build.sbt`

fommil commented on src/main/scala/net/cakesolutions/CakePublishPlugin.scala:
```diff
@@ -0,0 +1,264 @@
+  * Plugin for publishing artefacts (by default, configured for use by Cake
+  * Solution Limited artifacts).
+  */
+object CakePublishPlugin extends AutoPlugin {
```
`CakePublishMavenPlugin` ... publish typically means docker around here.

carlpulley commented:
For releasing the plugin to Sonatype, the release process will not alter. Code added here is simply functionality that is currently squirrelled away in sbt-sensible.

For projects that use this plugin (i.e. sbt-cake), they will use an alternative release strategy (as they'll be targeting non-Sonatype/non-Bintray repos).

Commit:
Name fixup to plugin

Commit:
Added in simple project plugin tests

Commit:
Corrected test project name

Commit:
Updated README

Commit:
Fix for git repos and CI environments

Commit:
Version bump on scalafmt

Commit:
README updates

Commit:
Removed sbt-sensible
Ensured code passes scalafmt tests

Commit:
Fixed library evictions

carlpulley commented:
Tag based releases work well on simple projects, however it breaks down on more realistic projects (e.g. when we need to version both endpoints and the main project code differently to each other - as is common in microservice architectures and all Play projects). This is a major pain point that we'll need to address in a follow up PR.

Commit:
Refactored based on arch. review feedback

Commit:
Use static SBT tasks (Def.task) instead of dynamic tasks (Def.taskDyn)
where possible

ferhtaydn reviewed (commented):
LGTM 👍

lashford commented on Jenkinsfile:
```diff
@@ -21,7 +21,7 @@ pipeline {
         ansiColor('xterm') {
           // TODO: CO-77: add in scalafmt::test target
-          // sh "sbt ';headerCheck ;scalastyle ;scalafmt'"
+          // sh "sbt ';headerCheck ;scalastyle ;scalafmt::test'"
```
why is this commented out?

carlpulley replied:
Originally, because we are unable to enable the scalafmt plugin within our project code. CO-77 speculates that this is a Scala version thing?

However, I've just performed a version bump and it appears that all is good with the newer version of the scalafmt plugin. I'll version bump this and enable these Jenkins build settings. I'll also mark CO-77 as no longer required.

lashford replied:
huzzah!  good things come to those who wait.

lashford commented on README.md:
````diff
@@ -213,3 +279,16 @@ being maintained and conflicting with [wartremover](https://github.com/wartremov
+git push --tags
+```
+
+In order for releasing to occur, individuals should have the Cake Solutions Sonatype PGP keys setup in their local
````
I don't agree that devs should be publishing releases, i would like only Jenkins to do the releasing.  What are thoughts regarding this?  Is this something that can be integrated into CI easily?

carlpulley replied:
Yes, we could easily add this to CI. However, we need to generate new Sonatype accounts, keys etc. just for Jenkins.

I'll add clarifying text that indicates that this situation is temporary.

I'll also generate a follow on Jira ticket for myself to chase up Sonatype regarding how we get others added to the main account.

lashford replied:
:+1:

carlpulley replied:
Created https://cakesolutions.jira.com/browse/CO-132 to track the Sonatype chase up work.

lashford commented on build.sbt:
```diff
@@ -40,19 +42,77 @@ addSbtPlugin("com.typesafe.sbt" % "sbt-digest" % "1.1.1")
 addSbtPlugin("com.typesafe.sbt" % "sbt-gzip" % "1.0.0")
 addSbtPlugin("com.typesafe.sbt" % "sbt-native-packager" % "1.1.6")
-addSbtPlugin("io.get-coursier" % "sbt-coursier" % "1.0.0-RC2")
+addSbtPlugin("io.get-coursier" % "sbt-coursier" % "1.0.0-RC3")
```
v1.0.0-RC6  is out, final release is coming soon.  Please add a todo in code to replace and new ticket to update in jira project.

carlpulley replied:
Based on comments above, I'll actually incorporate these changes into this PR.

Commit:
Merge branch 'master' into feature/co-32-release-strategy

Commit:
Version bumped scalafmt and enabled this plugin.
Other CR changes.

lashford commented on build.sbt:
```diff
@@ -42,15 +42,14 @@ addSbtPlugin("com.typesafe.sbt" % "sbt-digest" % "1.1.1")
 addSbtPlugin("com.typesafe.sbt" % "sbt-gzip" % "1.0.0")
 addSbtPlugin("com.typesafe.sbt" % "sbt-native-packager" % "1.1.6")
-addSbtPlugin("io.get-coursier" % "sbt-coursier" % "1.0.0-RC3")
+addSbtPlugin("io.get-coursier" % "sbt-coursier" % "1.0.0-RC5")
```
why not `RC6`

carlpulley replied:
I missed that release!

lashford commented on project/plugins.sbt:
```diff
@@ -16,6 +16,6 @@ addSbtPlugin("org.scalastyle" %% "scalastyle-sbt-plugin" % "0.8.0")
 // - see: https://github.com/lucidsoftware/neo-sbt-scalafmt
 // TODO: CO-77: load scalafmt plugin
-// addSbtPlugin("com.lucidchart" % "sbt-scalafmt" % "1.1")
+addSbtPlugin("com.lucidchart" % "sbt-scalafmt-coursier" % "1.7")
```
is this correct?  the comment about is reffering to scalafmt?

carlpulley replied:
As we are using coursier, it is now the correct thing to do. https://github.com/lucidsoftware/neo-sbt-scalafmt/blob/master/README.md#usage points out the importance of doing this in this type of use case.

Commit:
Version bump fix

lashford commented:
I think leaves us in a good place, i will verify the release process matches the process outlined in Cake's Confluence in terms of all the edge cases.  It looks very close so i will merge and make an additional comments in a new pr if there are any.
